x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/checkpoint/TimeBasedCheckpointProvider.java

```diff
@@ -96,8 +104,32 @@ public void createNextCheckpoint(final TransformCheckpoint lastCheckpoint, final
         );
     }
 
-    // for the purpose of testing
-    long getTime() {
-        return System.currentTimeMillis();
+    /**
+     * Aligns the timestamp with date histogram group source interval (if it is provided).
+     *
+     * @param timestamp timestamp to be aligned
+     * @param transformConfig transform configuration
+     * @return timestamp aligned with date histogram interval
+     */
+    private static long alignTimestamp(long timestamp, TransformConfig transformConfig) {
+        if (Boolean.FALSE.equals(transformConfig.getSettings().getInterimResults()) == false) {
+            return timestamp;
+        }
+        if (transformConfig.getPivotConfig() == null) {
+            return timestamp;
+        }
+        if (transformConfig.getPivotConfig().getGroupConfig() == null) {
+            return timestamp;
+        }
+        Map<String, SingleGroupSource> groups = transformConfig.getPivotConfig().getGroupConfig().getGroups();
+        if (groups == null || groups.isEmpty()) {
+            return timestamp;
+        }
+        Map.Entry<String, SingleGroupSource> topLevelGroupEntry = groups.entrySet().iterator().next();
+        if ((topLevelGroupEntry.getValue() instanceof DateHistogramGroupSource) == false) {
+            return timestamp;
+        }
```

Comment:
What the reason for requiring the top grouping?

Reply:
I was under impression that the date histogram grouping must be the first (top) specified in order to split all data into time buckets properly.
If it's not first then some other buckets (e.g.: terms) that are first will make the date histogram buckets mixed.
Does it make sense?

Reply:
Ok, nevermind.
I've changed it so that it searches for the first date histogram source that matches on field name.